### PR TITLE
Add support for png decoding on linux

### PR DESCRIPTION
**Pull request by r-zenine**
Hi @cpuhrsch

As discussed [here](https://github.com/pytorch/vision/pull/1632), This PR add's `c++` operators to decode `png`. 

Thanks

### CODE REVIEW

**Comment by r-zenine** [comment]
It's still a `WIP`, I have some weird issues with the search paths.

**Comment by r-zenine** [comment]
Hi @cpuhrsch, 

There first PR is done. The failures are due to pipelines timing out. I don't know how to fix that (already tried with and without ninja ). What could we do ? 

Thanks

**Comment by cpuhrsch** [comment]
Hello @r-zenine,

Thanks for splitting this out!

Master is currently timing out as well, I imagine it's because our builds are starting to take too long. Could you [try increasing the timeout](https://circleci.com/docs/2.0/configuration-reference/#example)? Especially with submodule libraries that are being built from source we're increasing the build times yet again. You could also build in parallel via the ```-j``` or ```--parallel``` flags of cmake. While Ninja causes an OOM, cmake might not.

On another note, let's make zlib a submodule as well. This way we're not increasing the number of dependencies. I thinks [this is the main repository](https://github.com/madler/zlib). Another comment on submodules: Since you can choose the commit it might make sense to pin it to a recent release for now, if you haven't done that already. That way we should run into less risk of a dependency bug.

- Christian

**Comment by r-zenine** [comment]
Hi, @cpuhrsch, 

Hope you are doing well!

I have added `zlib` as a vendored dependency and pinned the versions for both `libpng` and `zlib`. 

I increased the timeout for both `circleci` (30 minutes) and `travis` (1h). 

There is one last pipeline that is failing. The unit test for `ops` in failing on `windows`. I tried to rebase from `master` in the hope of getting a fix for that with no success. 

Have a good day! 

Let me know if there is anything else left to do!

**Comment by cpuhrsch** [comment]
That Windows build is a mystery to me. Paging Dr. Windows @peterjc123 to take a look if he has some time.

**Comment by peterjc123** [comment]
Well, the test failures are not related to this PR.

**Comment by cpuhrsch** [comment]
Ok, let's wait for the next nightly and try again.

**Comment by r-zenine** [comment]
Hi @cpuhrsch , 
Do you have the credentials to re-run the pipeline ?

File: test/test_image.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+def get_images(directory, img_ext):
+    assert os.path.isdir(directory)
+    for root, dir, files in os.walk(directory):
+        for fl in files:
+            _, ext = os.path.splitext(fl)
+            if ext == img_ext:
+                yield os.path.join(root, fl)
+
+
+class ImageTester(unittest.TestCase):
```
Comment (cpuhrsch) [inline test/test_image.py]: One thing that I missed: we should also exercise the edge cases. What happens if you pass an empty tensor, a Tensor that doesn't represent png, etc.

Comment (r-zenine) [reply test/test_image.py]: Good Catch,

I updated the test.

**Comment by r-zenine** [comment]
Hi @cpuhrsch, 

Please do not merge this PR. While working on the `jpeg` one. I figured out that even tough this approach works for now, it is not stricly what we wanted. 

Here is what is happening: 
We are building the extension code code against vendored versions of `libpng` and `zlib`. However, the code is running against the system wide dynamic libraries. It works because `libpng` and `zlib` and ubiquitous in linux systems. 

As in torch, we would need to ship the `.so` in package using `data_files` or `package_data` arguments in the `setup` function. My struggle now is that when I do this, the code fails because the dynamic library is not found. 

There is an option `runtime_library_dir` the might help. But there is little documentation about it. 
Do you happen to know how it's done in `pytorch` ? 

Thanks

File: setup.py [hunk 2]
```diff
@@ -196,6 +210,33 @@ def run(self):
...
+    this_dir = os.path.dirname(os.path.abspath(__file__))
+    if sys.platform.startswith('linux'):
+        cpu_count = multiprocessing.cpu_count()
+        os.chdir("third_party/zlib/")
+        throw_of_failure('cmake .')
+        throw_of_failure("cmake --build . --parallel {}".format(cpu_count))
+        os.chdir(this_dir)
+
+        zlib_path = os.path.join(this_dir, "third_party/zlib")
+        libpng_cmake_options="-DPNG_BUILD_ZLIB=ON -DPNG_STATIC=OFF -DZLIB_INCLUDE_DIR:PATH={zlib_path} -DZLIB_LIBRARY:FILEPATH={zlib_path}/libz.so".format(zlib_path=zlib_path)
```
Comment (cpuhrsch) [inline setup.py]: Try setting "PNG_STATIC=ON". Also see [https://stackoverflow.com/questions/43772887/cmake-creating-a-static-library](https://stackoverflow.com/questions/43772887/cmake-creating-a-static-library) or related articles.

Comment (r-zenine) [reply setup.py]: Hi, 

I Already tried that. That was actually my preferred option.  I spent a lot of time trying to link statically without success. Whatever the combination of options I tried, linking always fails with obscure errors.  The problem is that to link statically, we would need to compile position independent code ( with the -fPIC option) which for these libs does not seem to be working correctly.  

I am close to figuring out how to ship the dynamic libraries with the package. I'll keep you posted.

Comment (cpuhrsch) [reply setup.py]: This is [an example of a statically built library](https://github.com/pytorch/pytorch/blob/bc2e6d10fa7dc73b989ab3e22126867b50fed39c/aten/src/ATen/cpu/tbb/CMakeLists.txt) in PyTorch. You can use the [LD_DEBUG](http://man7.org/linux/man-pages/man8/ld.so.8.html) environment variable to print more information and see what libraries you're linking against so you can confirm the system shared libraries aren't being used. You can also use [the version macros](https://github.com/glennrp/libpng/blob/216387f6f7ee043df24a84ab2dc53d2827148b5e/CMakeLists.txt#L27) for the built libraries as another point of verification, assuming they're differing from the system library (I'd assume that the system libraries are older).

Another thing to add on top of this is uninstalling the linked libraries from the CI, but since they're so essential this might cause all kinds of unintended removals.

Comment (cpuhrsch) [reply setup.py]: We actually don't ship dynamic libraries anymore, because they can lead to all kinds of nightmares / conflicts with a given system.

Comment (r-zenine) [reply setup.py]: Could you please elaborate on the dynamic libraries part ? 
I just ran `ldd` on the `_C.so` extension for the torch nightly with the following output on my virtualenv.
```
	linux-vdso.so.1 (0x00007ffe4f261000)
	libshm.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libshm.so (0x00007ff98e5f9000)
	libtorch_python.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libtorch_python.so (0x00007ff986e26000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007ff986a9d000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ff9866ff000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007ff9864e7000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007ff9862c8000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ff985ed7000)
	libtorch.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libtorch.so (0x00007ff985cc3000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007ff985abb000)
	libtorch_cpu.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libtorch_cpu.so (0x00007ff976a5b000)
	libtorch_cuda.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libtorch_cuda.so (0x00007ff937be5000)
	libnvToolsExt-3965bdd0.so.1 => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libnvToolsExt-3965bdd0.so.1 (0x00007ff9379db000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007ff9377d7000)
	libc10_cuda.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libc10_cuda.so (0x00007ff9375a8000)
	libc10.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libc10.so (0x00007ff93734a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007ff98ea1e000)
	libgomp-7c85b1e2.so.1 => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libgomp-7c85b1e2.so.1 (0x00007ff937120000)
	libcudart-1b201d85.so.10.1 => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libcudart-1b201d85.so.10.1 (0x00007ff936ea1000)
```
Here is what the `lib` folder in the `torch` nightly package contains : 
```  
libc10.so       
libcaffe2_detectron_ops_gpu.so    
libcaffe2_nvrtc.so      
libcudart-1b201d85.so.10.1  
libnvrtc-5e8a26c9.so.10.1  
libnvToolsExt-3965bdd0.so.1  libtorch.so      
libtorch_cuda.so         
libtorch_python.so
libc10_cuda.so  
libcaffe2_module_test_dynamic.so  
libcaffe2_observers.so  
libgomp-7c85b1e2.so.1       
libnvrtc-builtins.so       
libshm.so                    
libtorch_cpu.so  
libtorch_global_deps.so 
```
These are all dynamic libraries shipped with `torch`. 

I have just figured out how to do the same while working on the `jpeg` library. I intend to do this here. Would you be okay with it ?

**Comment by r-zenine** [comment]
Hi @cpuhrsch , 

I have pushed a version with `libpng` and `zlib` linked dynamically and shipped with the package. 
The `windows` pipeline is failing without any useful error message. I will try to investigate further. 

What do you think of the approach ? Do we still try to link everything statically?

have a good day !

**Comment by cpuhrsch** [comment]
Hello @r-zenine,

Yes, ideally we link everything statically. Shipping dynamic libraries can be a nightmare and in particular building against a library that has dynamic dependencies can lead to lots of conflicts. If we can build something statically it's always preferred and if it seems possible we should try until it works.

Thanks,
Christian

**Comment by r-zenine** [comment]
hi @cpuhrsch , 

It works ! We are linking everything statically !
Thanks for pushing hard for this ! 

The only thing that remains is the windows error that has nothing to do with the PR ! 

Have a good day

**Comment by r-zenine** [comment]
Hi @cpuhrsch , 

What are the next steps ? Anything else i have to do ? 

Have a good day !

**Comment by cpuhrsch** [comment]
Great! Thanks again for your work! I'll take a closer look at the Windows build and also a few other things related to this PR setup. This might take a few days / week.

**Comment by r-zenine** [comment]
Hi @cpuhrsch , 

Let me work on MacOs support on the same PR. I'll try to push that on thursday. I will let you handle the windows pipeline and will get back to work on the `png` support. 

Thanks a lot for the help !

**Comment by r-zenine** [comment]
Hi @cpuhrsch ,

I just added `macos` support in the build. Tests for `image` are passing everywhere but other pipelines are failing because of other tests. It's probably due to something that happened with pytorch `nightly`.

Have a good day !!

**Comment by r-zenine** [comment]
Hi @cpuhrsch , 

I think the conda pipelines are failing because of a linking issue. There is a weird warning in the build :
```
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): did not find - or even know where to look for: /lib64/libm.so.6
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): did not find - or even know where to look for: /lib64/libc.so.6
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): did not find - or even know where to look for: $RPATH/libtorch_python.so
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): did not find - or even know where to look for: /lib64/ld-linux-x86-64.so.2
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): did not find - or even know where to look for: /lib64/libpthread.so.0
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): Needed DSO lib/libgcc_s.so.1 found in [u'libgcc-ng']
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): .. but [u'libgcc-ng'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): did not find - or even know where to look for: $RPATH/libpng16.so.16
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): did not find - or even know where to look for: $RPATH/libtorch_cpu.so
   INFO (torchvision,lib/python3.5/site-packages/torchvision/_C.so): Needed DSO lib/libcudart.so.10.0 found in defaults::cudatoolkit-10.0.130-0
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): did not find - or even know where to look for: $RPATH/libc10_cuda.so
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): Needed DSO lib/libstdc++.so.6 found in [u'libstdcxx-ng']
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): .. but [u'libstdcxx-ng'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): did not find - or even know where to look for: $RPATH/libtorch.so
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): did not find - or even know where to look for: $RPATH/libtorch_cuda.so
WARNING (torchvision,lib/python3.5/site-packages/torchvision/_C.so): did not find - or even know where to look for: $RPATH/libc10.so
WARNING (torchvision): dso library package defaults::python-3.5.6-hc3d631a_0 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```

Do you have any idea where it is coming from ? Do we have that on `master`, or is it specific to this PR ? 

Thanks

**Comment by cpuhrsch** [comment]
Hello @r-zenine,

We'll need to wait on @fmassa for some insight here. He's currently working on a hi-pri deadline so this will take a bit longer than expected. 

Is this working locally for you? You could also try installing this within a docker container (if your environment allows you to set that up) to get local feedback on the linking behavior etc.

In particular, if you take a look at the CI jobs you can see that they pull in the "pytorch/conda-cuda" container. This could be worthwhile looking into so you can get a faster turnaround time on experiments.

Thanks,
Christian

**Comment by r-zenine** [comment]
Hi @fmassa , @cpuhrsch 

Here is a status of what's going on, I tried to statically link the libraries (it works with wheel not conda and I could not figure out a fix with conda despite dedicating many hours to it). I also tried dynamically linking against `libpng` and shipping the library with the package ( like what `pytorch` does for it's dependencies). It's seems to be working on most pipeline. There is a bug with the linking in macos conda that I still need to figure out. 

If we go that way, it might make updating the windows pipelines easier. 

Thoughts ?

**Comment by andfoy** [comment]
@r-zenine, thanks for your contribution, we've migrated your commits to https://github.com/pytorch/vision/pull/2301, so we will close this one
